### PR TITLE
Update bindblenderprogressivetilecallback.cpp

### PR DESCRIPTION
**Pull request by anikket786**
Updated immediate mode openGL, to modern openGL, using openGL shading language 3.30 core.

### CODE REVIEW

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 1]
```diff
@@ -172,19 +172,82 @@ namespace
                     glBindTexture(GL_TEXTURE_2D, 0);
                     m_updated_buffer = false;
                 }
-
-                glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
-
-                glEnable(GL_TEXTURE_2D);
+                
+                int m_vertex_shader = glCreateShader(GL_VERTEX_SHADER);
+                glShaderSource(m_vertex_shader, 1, &m_vertex_shader_source, nullptr);
```
[the remark this answers is not in the document]
Comment (anikket786) [reply src/appleseed.python/bindblenderprogressivetilecallback.cpp]: HI, I have made changes. I wanna know Is the vertices data passed as arguments in the function needs to be updated each time the function is called or no ?

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 2]
```diff
@@ -172,20 +174,106 @@ namespace
...
-                glBegin(GL_QUADS);
-                    glTexCoord2f(0.0f, 1.0f); glVertex3f(x    ,  y    , 0.0f);
-                    glTexCoord2f(1.0f, 1.0f); glVertex3f(x + w,  y    , 0.0f);
-                    glTexCoord2f(1.0f, 0.0f); glVertex3f(x + w,  y + h, 0.0f);
-                    glTexCoord2f(0.0f, 0.0f); glVertex3f(x    ,  y + h, 0.0f);
-                glEnd();
-                glBindTexture(GL_TEXTURE_2D, 0);
-                glDisable(GL_TEXTURE_2D);
-            }
+                if(!m_updated_data_buffer)
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Please follow our coding conventions detailed here: https://github.com/appleseedhq/appleseed/wiki/Coding-Philosophy-and-Guidelines

That means: 4 spaces for indentation, insert space after `if`, `for`, `while`, etc.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 3]
```diff
@@ -172,20 +174,106 @@ namespace
...
+                  glShaderSource(vertex_shader_id, 1, &vertex_shader_source, nullptr);
+                  glCompileShader(vertex_shader_id);
+                  // Check for vertex_shader compile errors.
+                  int success;
+                  char infoLog[512];
+                  glGetShaderiv(vertex_shader_id, GL_COMPILE_STATUS, &success);
+                  if (!success)
+                  {
+                      glGetShaderInfoLog(vertex_shader_id, 512, nullptr, infoLog);
+                      std::cout << "ERROR::SHADER::VERTEX::COMPILATION_FAILED\n" << infoLog << std::endl;
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Don't write to `std::cout`, that's meaningless from the point of view of an integration (i.e. our Blender plugin).

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 4]
```diff
@@ -172,20 +174,106 @@ namespace
...
+                  GLuint vertex_shader_id = glCreateShader(GL_VERTEX_SHADER);
+                  glShaderSource(vertex_shader_id, 1, &vertex_shader_source, nullptr);
+                  glCompileShader(vertex_shader_id);
+                  // Check for vertex_shader compile errors.
+                  int success;
+                  char infoLog[512];
+                  glGetShaderiv(vertex_shader_id, GL_COMPILE_STATUS, &success);
+                  if (!success)
+                  {
+                      glGetShaderInfoLog(vertex_shader_id, 512, nullptr, infoLog);
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Don't repeat the magic 512 value. Make it a constant.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 5]
```diff
@@ -172,20 +174,106 @@ namespace
...
+                
+                  GLuint fragment_shader_id = glCreateShader(GL_FRAGMENT_SHADER);
+                  glShaderSource(fragment_shader_id, 1, &fragment_shader_source, nullptr);
+                  glCompileShader(fragment_shader_id);
+                  // Check for fragment_shader compile errors.
+                  glGetShaderiv(fragment_shader_id, GL_COMPILE_STATUS, &success);
+                  if (!success)
+                  {
+                      glGetShaderInfoLog(fragment_shader_id, 512, NULL, infoLog);
+                      std::cout << "ERROR::SHADER::FRAGMENT::COMPILATION_FAILED\n" << infoLog << std::endl;
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Remove.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 6]
```diff
@@ -172,20 +174,106 @@ namespace
...
+                
+                  m_shader_program_id = glCreateProgram();
+                  glAttachShader(m_shader_program_id, vertex_shader_id);
+                  glAttachShader(m_shader_program_id, fragment_shader_id);
+                  glLinkProgram(m_shader_program_id);
+                  // Check for linking errors.
+                  glGetProgramiv(m_shader_program_id, GL_LINK_STATUS, &success);
+                  if (!success) {
+                  glGetProgramInfoLog(m_shader_program_id, 512, NULL, infoLog);
+                  std::cout << "ERROR::SHADER::PROGRAM::LINKING_FAILED\n" << infoLog << std::endl;
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Remove.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 7]
```diff
@@ -172,20 +174,106 @@ namespace
...
+                  }
+                
+                  m_shader_program_id = glCreateProgram();
+                  glAttachShader(m_shader_program_id, vertex_shader_id);
+                  glAttachShader(m_shader_program_id, fragment_shader_id);
+                  glLinkProgram(m_shader_program_id);
+                  // Check for linking errors.
+                  glGetProgramiv(m_shader_program_id, GL_LINK_STATUS, &success);
+                  if (!success) {
+                  glGetProgramInfoLog(m_shader_program_id, 512, NULL, infoLog);
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Indent.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 8]
```diff
@@ -172,20 +174,106 @@ namespace
...
+                  {
+                      glGetShaderInfoLog(fragment_shader_id, 512, NULL, infoLog);
+                      std::cout << "ERROR::SHADER::FRAGMENT::COMPILATION_FAILED\n" << infoLog << std::endl;
+                  }
+                
+                  m_shader_program_id = glCreateProgram();
+                  glAttachShader(m_shader_program_id, vertex_shader_id);
+                  glAttachShader(m_shader_program_id, fragment_shader_id);
+                  glLinkProgram(m_shader_program_id);
+                  // Check for linking errors.
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Leave a blank line before full-line comments.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 9]
```diff
@@ -172,20 +174,106 @@ namespace
...
+                  glGetProgramiv(m_shader_program_id, GL_LINK_STATUS, &success);
+                  if (!success) {
+                  glGetProgramInfoLog(m_shader_program_id, 512, NULL, infoLog);
+                  std::cout << "ERROR::SHADER::PROGRAM::LINKING_FAILED\n" << infoLog << std::endl;
+                  }
+                  glDeleteShader(vertex_shader_id);
+                  glDeleteShader(fragment_shader_id);
+               
+                  //Vertex and texture coordinates.
+                  GLfloat vertices[] = {
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Move braces to their own lines.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 10]
```diff
@@ -172,20 +174,106 @@ namespace
...
+                  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, m_ebo_id);
+                  glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(indices), indices, GL_STATIC_DRAW);
+           
+                  glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 5 * sizeof(float), (void*)0);
+                  glEnableVertexAttribArray(0);
+                  glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 5 * sizeof(float), (void*)(3 * sizeof(float)));
+                  glEnableVertexAttribArray(1);
+                  glBindVertexArray(0);
+                  m_updated_data_buffer = true;
+               }
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Indentation is unclear/broken.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 11]
```diff
@@ -199,7 +287,13 @@ namespace
         size_t              m_texture_height;
 
         bpy::object         m_request_redraw_callback;
-
+        
+        GLuint              m_shader_program_id;
+        GLuint              m_vao_id, 
+                            m_vbo_id,
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Repeat the type.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 12]
```diff
@@ -211,6 +305,17 @@ namespace
                 m_texture_height = 0;
             }
         }
+        
+        void delete_shader_buffer()
+        {
+            if(m_updated_data_buffer)
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Insert space.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 13]
```diff
@@ -172,20 +174,106 @@ namespace
...
-                    glTexCoord2f(1.0f, 1.0f); glVertex3f(x + w,  y    , 0.0f);
-                    glTexCoord2f(1.0f, 0.0f); glVertex3f(x + w,  y + h, 0.0f);
-                    glTexCoord2f(0.0f, 0.0f); glVertex3f(x    ,  y + h, 0.0f);
-                glEnd();
-                glBindTexture(GL_TEXTURE_2D, 0);
-                glDisable(GL_TEXTURE_2D);
-            }
+                if(!m_updated_data_buffer)
+                {
+                  const char *vertex_shader_source  = "#version 330 core\n"
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Move star to the type's side.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 14]
```diff
@@ -172,20 +174,106 @@ namespace
...
+                                                      "layout (location = 0) in vec3 pos;\n"
+                                                      "layout (location = 1) in vec3 tex_coords;\n"
+                                                      "out texture_coords;\n"
+                                                      "void main()\n"
+                                                      "{\n"
+                                                      "   gl_Position = vec4(pos, 1.0);\n"
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Indent shader code the same way as C++ code.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp (hunk 3 above)
Comment (anikket786) [reply src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Should I leave it without any error message or is there any method to show error output from the point of view of an integration

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp (hunk 2 above)
Comment (dictoon) [reply src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Not resolved :)

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp (hunk 12 above)
Comment (dictoon) [reply src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Not resolved.

**Comment by est77** [comment]
I just realized that we don't need the shaders. The vertex and fragment shaders should have already been setup by Blender when calling the tile callback. In particular, we don't want to replace the fragment shader used by Blender, it contains the color management transform to display linear RGB images on the screen.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 15]
```diff
@@ -172,20 +174,112 @@ namespace
...
+                    }
+                    
+                    glDeleteShader(vertex_shader_id);
+                    glDeleteShader(fragment_shader_id);
+               
+                    //Vertex and texture coordinates.
+                    GLfloat vertices[] = 
+                    {
+                        x    , y    , 0.0f, 0.0f, 1.0f
+                      , x + w, y    , 0.0f, 1.0f, 1.0f
```
Comment (Narann) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Let's say I call `draw_pixels(A, B, C, D)` and `m_updated_data_buffer` is `false`. This mean the GLSL program is compiled and vertices are setted. `m_updated_data_buffer` is now `true`.

Then I call `draw_pixels(E, F, G, H)`. If I follow the code behavior correctly, buffer datas will not be update.

I suggest to separate the compilation GLSL code that can be done only once. And the  buffer update that is required at each `draw_pixels()` call.

Comment (anikket786) [reply src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Thanks, I was not sure earlier if the vertex data will be updated on each call of the draw_pixels() function or not. Thanks for suggestion 👍

Comment (anikket786) [reply src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Well this code should work without any problem

**Comment by anikket786** [comment]
>           I just realized that we don't need the shaders. The vertex and fragment shaders should have already been setup by Blender when calling the tile callback. In particular, we don't want to replace the fragment shader used by Blender, it contains the color management transform to display linear RGB images on the screen.

if shaders are already  been setup by blender then we need the appropriate shader Program id setup by blender to bind that shader program for passing the texture and  vertices.

**Comment by mororo250** [comment]
wouldn't glGetIntegerv(GL_CURRENT_PROGRAM, m_shader_program_id) fix this?

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 16]
```diff
@@ -172,20 +179,129 @@ namespace
...
+                        "   tex = vec4(texture, texture_coords);\n"
+                        "}\n\0";
+                    
+                    GLuint vertex_shader_id = glCreateShader(GL_VERTEX_SHADER);
+                    glShaderSource(vertex_shader_id, 1, &m_vertex_shader_source, nullptr);
+                    glCompileShader(vertex_shader_id);
+                    
+                    // Check for vertex_shader compile errors.
+                    int success;
+                    const int length = 512;
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Use `GLsizei` seems better here.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 17]
```diff
@@ -172,20 +179,129 @@ namespace
...
+                    m_vertex_shader_source  = 
+                        "#version 330 core\n"
+                        "layout (location = 0) in vec3 tex_coords;\n"
+                        "layout (location = 1) in vec3 pos;\n"
+                        "out texture_coords;\n"
+                        "void main()\n"
+                        "{\n"
+                        "   gl_Position = vec4(pos, 1.0);\n"
+                        "   texture_coords = vec3(tex_coords);\n
+                        "}\0";
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Why are `\0` necessary? C strings are always terminated by a null byte.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 18]
```diff
@@ -172,20 +179,129 @@ namespace
...
+                        "{\n"
+                        "   tex = vec4(texture, texture_coords);\n"
+                        "}\n\0";
+                    
+                    GLuint vertex_shader_id = glCreateShader(GL_VERTEX_SHADER);
+                    glShaderSource(vertex_shader_id, 1, &m_vertex_shader_source, nullptr);
+                    glCompileShader(vertex_shader_id);
+                    
+                    // Check for vertex_shader compile errors.
+                    int success;
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: `GLint`?

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 19]
```diff
@@ -172,20 +179,129 @@ namespace
...
+                        glGetProgramInfoLog(m_shader_program_id, length, nullptr, infoLog);
+                    }
+                    
+                    glDeleteShader(vertex_shader_id);
+                    glDeleteShader(fragment_shader_id);
+                    
+                    m_texture_coords[8] =
+                    {
+                        0.0f, 1.0f
+                      , 1.0f, 1.0f
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Move commas at the end of previous lines.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 20]
```diff
@@ -172,20 +179,129 @@ namespace
...
+                      , 1.0f, 0.0f
+                      , 0.0f, 0.0f
+                    };
+                    
+                    m_indices[6] =
+                    {   
+                        0, 1, 3
+                      , 1, 2, 3
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Same here.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 21]
```diff
@@ -172,20 +179,129 @@ namespace
...
+                    
+                    glBindVertexArray(0);
+                    m_updated_data_buffer = true;
+                }   
+                    
+                //Vertex coordinates.
+                m_vertex_coords[12] = 
+                {
+                    x    , y    , 0.0f
+                  , x + w, y    , 0.0f
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Same here.

File: src/appleseed.python/bindblenderprogressivetilecallback.cpp [hunk 22]
```diff
@@ -172,20 +179,129 @@ namespace
...
+                    glBindBuffer(GL_ARRAY_BUFFER, m_texture_vbo_id);
+                    glBufferData(GL_ARRAY_BUFFER, sizeof(m_texture_coords), m_texture_coords, GL_STATIC_DRAW);
+                    glVertexAttribPointer(0, 2, GL_FLOAT, GL_FALSE, 2 * sizeof(float), (void*)0);
+                    glEnableVertexAttribArray(0);
+                    
+                    glBindVertexArray(0);
+                    m_updated_data_buffer = true;
+                }   
+                    
+                //Vertex coordinates.
```
Comment (dictoon) [inline src/appleseed.python/bindblenderprogressivetilecallback.cpp]: Insert space after `//`.

**Comment by dictoon** [comment]
Work picked up by @mororo250 and PR superseded by #2549.

**Pull request closed without merging**